lib/web/terminal.go

```diff
@@ -90,23 +90,29 @@ type AuthProvider interface {
 
 // NewTerminal creates a web-based terminal based on WebSockets and returns a
 // new TerminalHandler.
-func NewTerminal(ctx context.Context, req TerminalRequest, authProvider AuthProvider, sessCtx *SessionContext) (*TerminalHandler, error) {
+func NewTerminal(ctx context.Context, req TerminalRequest, authProvider AuthProvider, sessCtx *SessionContext, servers []types.Server, sessionData session.Session) (*TerminalHandler, error) {
 	// Make sure whatever session is requested is a valid session.
-	_, err := session.ParseID(string(req.SessionID))
+	_, err := session.ParseID(string(sessionData.ID))
 	if err != nil {
 		return nil, trace.BadParameter("sid: invalid session id")
 	}
 
 	if req.Login == "" {
-		return nil, trace.BadParameter("login: missing login")
+		if sessionData.Login == "" {
+			return nil, trace.BadParameter("login: missing login")
+		}
+		req.Login = sessionData.Login
 	}
-	if req.Term.W <= 0 || req.Term.H <= 0 {
-		return nil, trace.BadParameter("term: bad term dimensions")
+
+	if req.Server == "" {
+		if sessionData.ServerID == "" {
+			return nil, trace.BadParameter("server: missing server")
+		}
+		req.Server = sessionData.ServerID
 	}
 
-	servers, err := authProvider.GetNodes(ctx, req.Namespace)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	if req.Term.W <= 0 || req.Term.H <= 0 {
```

Comment:
Should we check for a max value as well?

Reply:
This was just moved from further up, but I'm not sure what a reasonable max would be?

Reply:
As discussed I added a max check here.